src/graph_generator.cc

```diff
@@ -490,71 +591,49 @@ namespace detail {
-			if(prev_horizon != nullptr) {
-				// update "buffer_last_writer" and "last_collective_commands" structures to subsume pre-horizon commands
-				const auto prev_hid = prev_horizon->get_cid();
-				auto& this_node_data = node_data.at(node);
-				for(auto& blw_pair : this_node_data.buffer_last_writer) {
-					blw_pair.second.apply_to_values([prev_hid](std::optional<command_id> cid) -> std::optional<command_id> {
-						if(!cid) return cid;
-						return {std::max(prev_hid, *cid)};
-					});
-				}
-				for(auto& [cgid, cid] : this_node_data.last_collective_commands) {
-					cid = std::max(prev_hid, cid);
-				}
-				for(auto& [cgid, cid] : this_node_data.host_object_last_effects) {
-					cid = std::max(prev_hid, cid);
-				}
-				// update lowest previous horizon id (for later command deletion)
-				if(lowest_prev_hid == 0) {
-					lowest_prev_hid = prev_hid;
-				} else {
-					lowest_prev_hid = std::min(lowest_prev_hid, prev_hid);
-				}
-
-				// We also use the previous horizon as the new init cmd for host-initialized buffers
-				node_data[node].current_init_cid = prev_hid;
-			}
+	void graph_generator::generate_epoch_dependencies(abstract_command* cmd) {
+		// No command must be re-ordered before its last preceding epoch to enforce the barrier semantics of epochs.
+		// To guarantee that each node has a transitive true dependency (=temporal dependency) on the epoch, it is enough to add an epoch -> command dependency
+		// to any command that has no other true dependencies itself and no graph traversal is necessary. This can be verified by a simple induction proof.
+
+		// As long the first epoch is present in the graph, all transitive dependencies will be visible and the initial epoch commands (tid 0) are the only
+		// commands with no true predecessor. As soon as the first epoch is pruned through the horizon mechanism however, more than one node with no true
+		// predecessor can appear (when visualizing the graph). This does not violate the ordering constraint however, because all "free-floating" nodes
+		// in that snapshot had a true-dependency chain to their predecessor epoch at the point they were flushed, which is sufficient for following the
+		// dependency chain from the executor perspective.
+
+		if(const auto deps = cmd->get_dependencies();
+		    std::none_of(deps.begin(), deps.end(), [](const abstract_command::dependency d) { return d.kind == dependency_kind::TRUE_DEP; })) {
+			auto last_epoch = node_data.at(cmd->get_nid()).epoch_for_new_commands;
+			assert(cmd->get_cid() != last_epoch);
+			cdag.add_dependency(cmd, cdag.get(last_epoch), dependency_kind::TRUE_DEP, dependency_origin::last_epoch);
 		}
-		// After updating all the data structures, delete before-cleanup-horizon commands
-		// Also remove commands from command_buffer_reads (if it exists)
-		if(cleanup_horizon_id > 0) {
+	}
+
+	void graph_generator::prune_commands_before(const command_id min_epoch) {
+		if(min_epoch > min_epoch_last_pruned_before) {
```

Comment:
Same as in `task_manager`, may be clearer with `!=`?

Reply:
Same objection, the `>` highlights the expected monotone progress.